src/MOI_wrapper/MOI_nonlinear.jl

```diff
@@ -0,0 +1,349 @@
+            )
+        elseif typeof(s) == MOI.ScalarAffineFunction{Float64}
+            if length(s.terms) > 1
+                append!(opcode, GRB_OPCODE_PLUS)
+                append!(data, -1.0)
+                append!(parent, parent_index)
+                parent_index += Cint(1)
+            end
+            if s.constant != 0.0
+                append!(opcode, GRB_OPCODE_CONSTANT)
+                append!(data, s.constant)
+                append!(parent, parent_index)
+                current_index += Cint(1)
+            end
+            for term in s.terms
+                var_index = c_column(model, term.variable)
+                coeff = term.coefficient
+                current_index = _add_expression_tree_variable(
+                    opcode,
+                    data,
+                    parent,
+                    coeff,
+                    var_index,
+                    current_index,
+                    parent_index,
+                )
+                current_index += Cint(1)
+            end
+        elseif typeof(s) == Float64
+            _add_expression_tree_constant(opcode, data, parent, s, parent_index)
+        elseif typeof(s) == Int64
+            _add_expression_tree_constant(
+                opcode,
+                data,
+                parent,
+                convert(Float64, s),
+                parent_index,
+            )
+        else
+            throw(MOI.UnsupportedAttribute(s))
+        end
+    end
+
+    return
+end
+
+function MOI.add_constraint(
+    model::Optimizer,
+    f::MOI.ScalarNonlinearFunction,
+    s::_SCALAR_SETS,
+)
+    opcode = Vector{Cint}()
+    data = Vector{Cdouble}()
+    parent = Vector{Cint}()
+
+    sense, rhs = _sense_and_rhs(s)
+
+    _process_nonlinear(model, f, opcode, data, parent)
+
+    # Add resultant variable
+    vi, ci = MOI.add_constrained_variable(model, s)
+    resvar_index = c_column(model, vi)
+
+    ret = GRBaddgenconstrNL(
+        model,
+        C_NULL,
+        resvar_index,
+        length(opcode),
+        opcode,
+        data,
+        parent,
+    )
+    _check_ret(model, ret)
+    # GRBwrite(model, "checkjl.lp")
+    _require_update(model, model_change = true)
+    model.last_constraint_index += 1
+    model.nl_constraint_info[model.last_constraint_index] =
+        _NLConstraintInfo(length(model.nl_constraint_info) + 1, s, vi)
+    return MOI.ConstraintIndex{MOI.ScalarNonlinearFunction,typeof(s)}(
+        model.last_constraint_index,
+    )
+end
+
+function MOI.is_valid(
+    model::Optimizer,
+    c::MOI.ConstraintIndex{MOI.ScalarNonlinearFunction,S},
+) where {S}
+    info = get(model.nl_constraint_info, c.value, nothing)
+    return info !== nothing && typeof(info.set) == S
+end
+
+function MOI.delete(
+    model::Optimizer,
+    c::MOI.ConstraintIndex{MOI.ScalarNonlinearFunction,S},
+) where {S}
+    info = _info(model, c)
+    _update_if_necessary(model)
+
+    ret = GRBdelgenconstrs(model, 1, [Cint(info.row - 1)])
```

Comment:
```suggestion
    ret = GRBdelgenconstrs(model, 1, Ref(Cint(info.row - 1)))
```

`[]` allocates a new vector, but really we just need to pass a pointer to something with one element.